In [27]:
import pandas as pd
import numpy as np

In [28]:
#download the class movie reviews from here:
#https://docs.google.com/spreadsheets/d/17rCJzmWxqvAu9rkpkgt4ToccIlY4A1Ffuu1W9X3B8Ag/
#then read that in as a pandas dataframe

In [29]:
M = pd.read_csv("Movies.csv", index_col='Name')

In [30]:
M.head()

American Sniper  The Hunger Games: Mockingjay - Part 1  \
Name                                                                         
Aarti Jaiswal                   4.0                                    NaN   
Aditya Dharmasagar              5.0                                    NaN   
Adrian Cavallaris               NaN                                    4.0   
Ahmed Muheebuddin               5.0                                    NaN   
Aishwarya reddy                 NaN                                    3.0   

                    Guardians of the Galaxy  The Lego Movie  The Hobbit  \
Name                                                                      
Aarti Jaiswal                           NaN             NaN         4.0   
Aditya Dharmasagar                      NaN             4.0         5.0   
Adrian Cavallaris                       3.0             NaN         4.0   
Ahmed Muheebuddin                       4.0             NaN         3.0   
Aishwarya reddy                         4.0             3.0         4.0   

                    Transformers  Malificent  Big Hero 6  Godzilla  \
Name                                                                 
Aarti Jaiswal                3.0         NaN         5.0       NaN   
Aditya Dharmasagar           4.0         4.0         NaN       4.0   
Adrian Cavallaris            2.0         NaN         NaN       NaN   
Ahmed Muheebuddin            NaN         NaN         NaN       NaN   
Aishwarya reddy              4.0         3.0         NaN       4.0   

                    Interstellar  How to Train your Dragon 2  Gone Girl  \
Name                                                                      
Aarti Jaiswal                NaN                         NaN        4.0   
Aditya Dharmasagar           5.0                         5.0        5.0   
Adrian Cavallaris            NaN                         NaN        NaN   
Ahmed Muheebuddin            NaN                         NaN        NaN   
Aishwarya reddy              4.0                         3.0        5.0   

                    Divergent  The Fault in Our Stars  Unbroken  \
Name                                                              
Aarti Jaiswal             NaN                     NaN       NaN   
Aditya Dharmasagar        4.0                     4.0       NaN   
Adrian Cavallaris         4.0                     NaN       NaN   
Ahmed Muheebuddin         NaN                     NaN       NaN   
Aishwarya reddy           NaN                     4.0       NaN   

                    300: Rise of an Empire  
Name                                        
Aarti Jaiswal                          NaN  
Aditya Dharmasagar                     5.0  
Adrian Cavallaris                      NaN  
Ahmed Muheebuddin                      NaN  
Aishwarya reddy                        3.0

In [31]:
def pearson(s1, s2):
    """Take two pd.Series objects and return a pearson correlation."""
    s1_c = s1 - s1.mean()
    s2_c = s2 - s2.mean()
    return np.sum(s1_c * s2_c) / np.sqrt(np.sum(s1_c ** 2) * np.sum(s2_c ** 2))

In [32]:
def get_recs(movie_name, M, num):

    import numpy as np
    reviews = []
    for title in M.columns:
        if title == movie_name:
            continue
        cor = pearson(M[movie_name], M[title])
        if np.isnan(cor):
            continue
        else:
            reviews.append((title, cor))
    
    reviews.sort(key=lambda tup: tup[1], reverse=True)
    return reviews[:num]

    

In [33]:
pearson(M['The Fault in Our Stars'], M['American Sniper'])

-0.12571904015492413

In [34]:
pearson(M['The Fault in Our Stars'], M['The Hunger Games: Mockingjay - Part 1'])

0.059150717691497816

In [35]:
pearson(M['The Fault in Our Stars'], M['Guardians of the Galaxy'])

0.14815495894813271

In [36]:
pearson(M['The Fault in Our Stars'], M['The Lego Movie'])

0.059899519128956448

In [37]:
pearson(M['The Fault in Our Stars'], M['The Hobbit'])

-0.0064260743251112282

In [38]:
pearson(M['The Fault in Our Stars'], M['Transformers'])

0.010345558567015564

In [39]:
pearson(M['The Fault in Our Stars'], M['Malificent'])

0.213754032137282

In [40]:
pearson(M['The Fault in Our Stars'], M['Big Hero 6'])

-0.056614041247155016

In [41]:
pearson(M['The Fault in Our Stars'], M['Godzilla'])

-0.047355933368302426

In [42]:
pearson(M['The Fault in Our Stars'], M['Interstellar'])

0.25035799093321581

In [43]:
pearson(M['The Fault in Our Stars'], M['How to Train your Dragon 2'])

0.15769576025447329

In [44]:
pearson(M['The Fault in Our Stars'], M['Gone Girl'])

0.069028758717867775

In [45]:
pearson(M['The Fault in Our Stars'], M['Divergent'])

0.17470709093502146

In [46]:
pearson(M['The Fault in Our Stars'], M['Unbroken'])

0.075201321328153076

In [47]:
pearson(M['The Fault in Our Stars'], M['300: Rise of an Empire'])

-0.046920362656838299

In [48]:
pearson(M['The Fault in Our Stars'], M['The Fault in Our Stars'])

1.0

In [49]:
recs = get_recs('The Fault in Our Stars', M, 10)

In [50]:
recs[:10]

[('Interstellar', 0.25035799093321581),
 ('Malificent', 0.213754032137282),
 ('Divergent', 0.17470709093502146),
 ('How to Train your Dragon 2', 0.15769576025447329),
 ('Guardians of the Galaxy', 0.14815495894813271),
 ('Unbroken', 0.075201321328153076),
 ('Gone Girl', 0.069028758717867775),
 ('The Lego Movie', 0.059899519128956448),
 ('The Hunger Games: Mockingjay - Part 1', 0.059150717691497816),
 ('Transformers', 0.010345558567015564)]

In [51]:
anti_recs = get_recs('The Fault in Our Stars', M, 8551)

In [52]:
anti_recs[:-10]

[('Interstellar', 0.25035799093321581),
 ('Malificent', 0.213754032137282),
 ('Divergent', 0.17470709093502146),
 ('How to Train your Dragon 2', 0.15769576025447329),
 ('Guardians of the Galaxy', 0.14815495894813271)]

**Question 1:  What movie is most Similar to 'The fault in our stars' (60 pts)**


According to Pearson's R, 'Interstellar' is the movie most similar to 'The Fault in Our Stars' relatively speaking.

**QUESTION 2**

In [53]:
#HINTS...

#the movies I've seen
my_ratings = M.ix['Sohaib Ahmed'].dropna()
#the movie I've rated the highest
my_ratings.sort(ascending=False)
#the name of the movie I've rated the highest
fav_movie = my_ratings.index[0]

/Applications/Python/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:6: FutureWarning: sort is deprecated, use sort_values(inplace=True) for INPLACE sorting


In [58]:
print my_ratings

Gone Girl       5.0
Interstellar    5.0
Transformers    4.0
The Hobbit      3.0
Name: Sohaib Ahmed, dtype: float64


In [59]:
print fav_movie

Gone Girl


In [63]:
#the movies the class has seen
class_ratings = M.dropna()
#the movie the class rated the highest
class_ratings.sort(ascending=False)

/Applications/Python/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:4: FutureWarning: sort(....) is deprecated, use sort_index(.....)


In [64]:
print class_ratings

                    American Sniper  The Hunger Games: Mockingjay - Part 1  \
Name                                                                         
Ashutosh Sureka                 5.0                                    4.0   
Benjamin Baysinger              5.0                                    4.0   
Bhuvana Posina                  5.0                                    4.0   
Rajesh Jaiswal                  4.0                                    3.0   
Shreya Kulkarni                 4.0                                    5.0   
Chaitanya Varma                 4.0                                    1.0   
Gaurav Karale                   5.0                                    4.0   
Rohit Ukirde                    5.0                                    4.0   

                    Guardians of the Galaxy  The Lego Movie  The Hobbit  \
Name                                                                      
Ashutosh Sureka                         5.0             5.0         5

In [65]:
recs[:10]

[('Interstellar', 0.25035799093321581),
 ('Malificent', 0.213754032137282),
 ('Divergent', 0.17470709093502146),
 ('How to Train your Dragon 2', 0.15769576025447329),
 ('Guardians of the Galaxy', 0.14815495894813271),
 ('Unbroken', 0.075201321328153076),
 ('Gone Girl', 0.069028758717867775),
 ('The Lego Movie', 0.059899519128956448),
 ('The Hunger Games: Mockingjay - Part 1', 0.059150717691497816),
 ('Transformers', 0.010345558567015564)]

In [66]:
anti_recs = get_recs('Gone Girl', M, 8551)

In [67]:
anti_recs[:-10]

[('How to Train your Dragon 2', 0.20015230349034269),
 ('American Sniper', 0.19992105616147593),
 ('The Lego Movie', 0.1183046729317008),
 ('Godzilla', 0.11432590070503264),
 ('Unbroken', 0.077333281370714532)]

**Question 2:  Which movie(s) would you most like to see, based on your classmates experience? (40 pts)**

Based on Pearson's R and based on the class's ratings, Since I have already seen Interstellar, Transformers, and The Hobbit, I would want to watch the Malificent next based on my classmates' experience:

 ('Malificent', 0.213754032137282),
 ('Divergent', 0.17470709093502146),
 ('How to Train your Dragon 2', 0.15769576025447329),
 ('Guardians of the Galaxy', 0.14815495894813271),
 ('Unbroken', 0.075201321328153076),
 ('The Lego Movie', 0.059899519128956448),
 ('The Hunger Games: Mockingjay - Part 1', 0.059150717691497816),

###Question 3: Bonus Question...  For all the movies you haven't seen, can you predict how you'd rate them using your the class reviews? (10 pts)

**Can you provide the solution to Q3? Thank you!**